src/ppm.rs

```diff
@@ -0,0 +1,473 @@
+//!   |                      |    aggregate request    |
+//!   |                      |------------------------>| prio_verify_start()
+//!   |                      |                         |
+//!   |                      |                         |
+//!   |                      |<------------------------|
+//!   |                      | prio_verify_finish()    | prio_verify_finish()
+//!   v                      v                         v
+//!                        output share              output share
+//! ```
+//!
+//! NOTE: This protocol implemented here is a prototype and has not undergone security analysis.
+//! Use at your own risk.
+
+use std::convert::TryFrom;
+
+use getrandom::getrandom;
+
+use crate::field::{rand, FieldElement};
+use crate::pcp::{decide, prove, query, PcpError, Proof, Value, Verifier};
+use crate::prng::{Prng, PrngError, SEED_LENGTH};
+
+/// Errors emitted by this module.
+#[derive(Debug, thiserror::Error)]
+pub enum PpmError {
+    /// An error occurred.
+    #[error("ppm error: {0}")]
+    Ppm(String),
+
+    /// The distributed input was deemed invalid.
+    #[error("ppm error: invalid distributed input: {0}")]
+    Validity(&'static str),
+
+    /// PCP error.
+    #[error("ppm error: pcp error: {0}")]
+    Pcp(#[from] PcpError),
+
+    /// PRNG error.
+    #[error("ppm error: prng error: {0}")]
+    Prng(#[from] PrngError),
+
+    /// Calling get_random() returned an error.
+    #[error("ppm error: getrandom error: {0}")]
+    GetRandom(#[from] getrandom::Error),
+}
+
+/// A share of an input or proof for Prio.
+pub enum PrioShare<F: FieldElement> {
+    /// An uncompressed share, typically sent to the leader.
+    Leader(Vec<F>),
+
+    /// A compressed share, typically sent to the helper. The first parameter is a seed for a
+    /// pseudorandom generator
+    Helper(Vec<u8>, usize),
+}
+
+impl<F: FieldElement> TryFrom<PrioShare<F>> for Vec<F> {
+    type Error = PpmError;
+
+    fn try_from(share: PrioShare<F>) -> Result<Self, PpmError> {
+        match share {
+            PrioShare::Leader(data) => Ok(data),
+            PrioShare::Helper(seed, length) => {
+                let mut prng = Prng::from_seed(&seed)?;
+                Ok((0..length).map(|_| prng.next().unwrap()).collect())
+            }
+        }
+    }
+}
+
+/// The message sent by the client to each aggregator. This includes the client's input share and
+/// the initial message of the input-validation protocol.
+pub struct PrioUpload<F: FieldElement> {
+    /// The identity of the intended recipient.
+    pub aggregator_id: u8,
+
+    /// The input share.
+    pub input_share: PrioShare<F>,
+
+    /// The proof share.
+    pub proof_share: PrioShare<F>,
+
+    /// The sum of the joint randomness seed shares sent to the other aggregators.
+    pub joint_rand_seed_hint: Vec<u8>,
+
+    /// The blinding factor, used to derive the aggregator's joint randomness seed share.
+    pub blind: Vec<u8>,
```

Comment:
That would allow the client to choose *any* joint rand it wants, which violates soundness. Deriving joint rand from the input shares removes the client's influence on the distribution of the joint rand, at least when we model the hash function as a random oracle. The purpose of `blind`, at a very high level, is to prevent the joint rand shares from leaking information about the input shares to other aggregators.

Bare in mind that all of this needs a proof, but what we're following the lead of the Crypto 2019 paper. This technique is suggested, though not spelled out completely, in Section 6.

Reply:
I see what I was missing now -- in Prio v2, we trusted the ingestion servers to provide joint randomness, not the client, but we do not assume any such actor in PPM. This makes sense to me now, thank you!

Reply:
Yes, that's correct, thanks for the clarification.